Description: Rename get_service_credentials.py
Author: Thomas Goirand <zigo@debian.org>
Bug-Ubuntu: https://bugs.launchpad.net/fuel/+bug/1542818
Forwarded: no
Last-Update: 2016-02-07

--- fuel-library-9.0~0+2016.02.04.git.8094062c62+dfsg1.orig/deployment/puppet/docker/manifests/dockerctl.pp
+++ fuel-library-9.0~0+2016.02.04.git.8094062c62+dfsg1/deployment/puppet/docker/manifests/dockerctl.pp
@@ -27,7 +27,7 @@ class docker::dockerctl (
   #  group   => 'root',
   #  mode    => '0755',
   #}
-  #file { "$bin_dir/get_service_credentials.py":
+  #file { "$bin_dir/fuel_get_service_credentials":
   #  mode    => 0755,
   #  content => template("docker/get_service_credentials.py.erb")
   #}
--- fuel-library-9.0~0+2016.02.04.git.8094062c62+dfsg1.orig/files/fuel-docker-utils/functions.sh
+++ fuel-library-9.0~0+2016.02.04.git.8094062c62+dfsg1/files/fuel-docker-utils/functions.sh
@@ -124,7 +124,7 @@ function retry_checker {
 
 function get_service_credentials {
   credentialfile=$(mktemp /tmp/servicepws.XXXXX)
-  get_service_credentials.py $ASTUTE_YAML > $credentialfile
+  fuel_get_service_credentials $ASTUTE_YAML > $credentialfile
   . $credentialfile
   rm -f $credentialfile
 }
