--- a/manager.py        2012-10-31 14:09:36.153257689 +0400
+++ b/manager.py        2012-10-31 14:10:01.261260402 +0400
@@ -65,11 +65,11 @@
                       'internal_access_path': internal_access_path,
                       'last_activity_at': time.time()}
         data = jsonutils.dumps(token_dict)
-        self.mc.set(token, data, FLAGS.console_token_ttl)
+        self.mc.set(token.encode('UTF-8'), data, FLAGS.console_token_ttl)
         LOG.audit(_("Received Token: %(token)s, %(token_dict)s)"), locals())
 
     def check_token(self, context, token):
-        token_str = self.mc.get(token)
+        token_str = self.mc.get(token.encode('UTF-8'))
         token_valid = (token_str is not None)
         LOG.audit(_("Checking Token: %(token)s, %(token_valid)s)"), locals())
         if token_valid:

